### DISCUSSION TITLE
Edge Function only Invoking when Anon Policy is actvated for table we want to use the Service Role Key

### DISCUSSION THREAD

**Discussion by farmhutsoftwareteam** [supabase/supabase#23172] [Questions] [answered]
## Title
Issue with Supabase RLS Policy and Edge Function Key

## Description
I'm encountering an issue with the Row Level Security (RLS) policy in Supabase. The policy is set to only allow inserts with a service key. However, my edge function is using an anonymous key, which is causing the function to fail.

## Steps to Reproduce
1. Set RLS policy on a table to only allow inserts with a service key.
2. Create an edge function that performs an insert operation on the table using an anonymous key.
3. Run the edge function.

## Expected Behavior
The edge function should successfully insert data into the table.

## Actual Behavior
The edge function fails due to the RLS policy.It only works when the table allows anon insertes

## Additional Information
I'm considering modifying the edge function to use a service key instead of an anonymous key, but I'm aware this could have security implications as the service key bypasses all RLS policies and has full access to the database. I'm also considering creating a stored procedure that encapsulates the insert operation and grants the necessary permissions to the anonymous role to execute this procedure. Any advice or suggestions would be greatly appreciated.

**Comment by GaryAustin1** [comment, marked as answer]
The only way you can access a table thru the API and bypass RLS is with the service role key.   It is normal to use it in edge functions and I believe it is an environment variable for you to use.

You of course are responsible that your edge function does not do anything you don't want in your actual code then in the edge function.

As you commented you blocked anon with RLS (no anon policy) then added it and it works as you were using the anon key in the edge function.    Service role will not look at RLS at all so you can have no policies, or policies for anon and authenticated and it will still have access.  You don't need to set the TO option to include it as it won't look at the policies at all.
